gearpump/gearpump
# fix #1981, Support OAuth2 Social login

OAuth2 authentication PR for discussion.

**How to test**

1. enable OAuth2 by changing "gearpump.ui-security.oauth2-authenticator-enabled" to true in gear.conf
2. Set proxy before starting UI server, for example:
   ```
    export JAVA_OPTS=-Dhttp.proxyHost=xx.com -Dhttp.proxyPort=8088 -Dhttps.proxyHost=xx.com -Dhttps.proxyPort=8088
   ```
3. Start the UI server by ```bin\services```
4. It will then show the "social login" buttons.
![social](https://cloud.githubusercontent.com/assets/2595532/13786276/d8eb7c44-eb11-11e5-905f-5c9f225f5891.png)
5. Currently, only Google account and Cloudfoundry icons are showed. (NOTE: Cloudfoundry  not tested on real server yet)
6. Click the "Google" icon and see what happens.

**How to configure**
Each OAuth2Authenticator has a config like this:
```
"google" {
        "class" = "io.gearpump.services.security.oauth2.impl.GoogleOAuth2Authenticator"

        ## Replace "127.0.0.1:8090" with your actual project site root URL.
        "callback" = "http://127.0.0.1:8090/login/oauth2/google/callback"

        ## !!NOTE!! Replace clientID and clientSecret with your own Application to avoid
        ## potential privacy leakage.
        "clientId" = "xxxxxxxxxxxxxxxxxxx"
        "clientSecret" = "xxxx"

        ## Available values for default user role: guest, user, admin
        "defaultUserRole" = "guest"

        ## ICon for this OAuth2 provider
        icon = "/icons/google.png"
      }
```

**TODO**
1. Add unit test
2. Find a working cloudfoundry UAA server to test whether the UAA OAuth2 authentication works.
